src/libraries/Microsoft.PowerFx.Interpreter/EvalContext.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.PowerFx.Interpreter
+{
+    internal class EvalContext
+    {
+        private int _currentCallDepth = 0;
+
+        // We can set this up to read from powerfx config, but it isn't nessecary yet.
+        private readonly int _maxCallDepth;
+        private readonly Mutex _mutex = new Mutex();
+
+        internal EvalContext(int maxCallDepth = 100)
+        {
+            _maxCallDepth = maxCallDepth;
+        }
+
+        /// <summary>
+        /// Increments the runtime call depth counter.
+        /// Is threadsafe.
+        /// </summary>
+        /// <returns>False when you go over the recursive depth limit, true normally.</returns>
+        public bool IncrementCallDepth()
+        {
+            _mutex.WaitOne();
+            _currentCallDepth++;
+            if (_currentCallDepth >= _maxCallDepth)
```

Comment:
I don't think we need Mutex for this, Interlocked.Increment/Decrement should be sufficient: 
[https://docs.microsoft.com/en-us/dotnet/api/system.threading.interlocked.increment?view=net-6.0](https://docs.microsoft.com/en-us/dotnet/api/system.threading.interlocked.increment?view=net-6.0)

Reply:
I'm also not sure we need to be that careful about concurrent access here, at the moment EvalVistitor is single threaded.

Reply:
Changed to use Interlocked#Increment and Interlocked#Decrement